hazelcast/src/main/java/com/hazelcast/instance/impl/ClusterTopologyIntentTrackerImpl.java

```diff
@@ -0,0 +1,264 @@
+ * |                                                                 | STABLE(5)                       |
+ * +-----------------------------------------------------------------+---------------------------------+
+ * | $ kubectl delete pod hz-hazelcast-enterprise-2                  | MISSING_MEMBERS(5) ->           |
+ * |   (simulating kubernetes deleted a pod)                         | (after pod is restarted)        |
+ * |                                                                 | STABLE(5)                       |
+ * +-----------------------------------------------------------------+---------------------------------+
+ * | $ kubectl scale sts hz-hazelcast-enterprise --replicas 0        | CLUSTER_SHUTDOWN(0)             |
+ * +-----------------------------------------------------------------+---------------------------------+
+ * }
+ * </pre>
+ */
+public class ClusterTopologyIntentTrackerImpl implements ClusterTopologyIntentTracker {
+
+    /**
+     * Currently detected cluster topology intent.
+     */
+    private final AtomicReference<ClusterTopologyIntent> clusterTopologyIntent =
+            new AtomicReference<>(ClusterTopologyIntent.NOT_IN_MANAGED_CONTEXT);
+    // single-threaded executor for actions in response to cluster topology intent changes
+    private final ExecutorService clusterTopologyExecutor;
+    // applies when automatic cluster state management is enabled (with persistence in kubernetes)
+    private final ClusterState clusterStateForMissingMembers;
+    private final ILogger logger;
+    private final Node node;
+    /**
+     * The desired number of members, as specified in the runtime environment. eg in kubernetes
+     * {@code kubectl scale sts hz --replicas 5} means {@code currentClusterSpecSize} is 5.
+     */
+    private volatile int currentClusterSpecSize = UNKNOWN;
+
+    public ClusterTopologyIntentTrackerImpl(Node node) {
+        this.clusterStateForMissingMembers = node.getProperties()
+                .getEnum(PERSISTENCE_AUTO_CLUSTER_STATE_STRATEGY, ClusterState.class);
+        if (clusterStateForMissingMembers != ClusterState.FROZEN
+                && clusterStateForMissingMembers != ClusterState.NO_MIGRATION) {
+            throw new InvalidConfigurationException("Value of property " + PERSISTENCE_AUTO_CLUSTER_STATE_STRATEGY.getName()
+                    + " was " + clusterStateForMissingMembers + " but should be one of FROZEN, NO_MIGRATION.");
+        }
+        this.clusterTopologyExecutor = Executors.newSingleThreadExecutor();
+        this.logger = node.getLogger(ClusterTopologyIntentTracker.class);
+        this.node = node;
+    }
+
+    public void shutdown() {
+        clusterTopologyExecutor.shutdown();
+    }
+
+    @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity"})
+    @Override
+    public void update(int previousClusterSpecSize, int currentClusterSpecSize, int readyNodesCount, int currentNodesCount) {
```

Comment:
maybe if we can make this method shorter by dividing it into sub-methods, it can be more easy understand.
like this: find intent and then do or not-do something...

Reply:
see 959ee605202d8b463600200c977bcddbb4796e47 and subsequent commits. Now the update to next intent is collocated with setting up a follow-up `Runnable` action and I think it is clearer.